tests/integration/opup_test.py

```diff
@@ -0,0 +1,166 @@
+            if pt.OpUpFeeSource.GroupCredit
+            else sp.fee
+        )
+
+        result = _dryrun(maximize_budget, sp, accounts)
+
+        assert result.passed()
+        assert result.budget_added() == _application_opcode_budget * inner_txn_count
+    else:
+        # Withholding account and/or transaction fee funding fails the
+        # transaction.
+        sp = DryRunExecutor.SUGGESTED_PARAMS
+        sp.flat_fee = True
+        sp.fee = algosdk.constants.min_txn_fee
+        result = _dryrun(maximize_budget, DryRunExecutor.SUGGESTED_PARAMS, [])
+        assert not result.passed()
+
+
+@pytest.mark.parametrize("source", [f for f in pt.OpUpFeeSource])
+@pytest.mark.parametrize("budget_added", list(range(1_000, 20_000, 2_500)))
+@pytest.mark.parametrize("with_funding", [True, False])
+def test_opup_ensure_budget(
+    source: pt.OpUpFeeSource, budget_added: int, with_funding: bool
+):
+    inner_txn_count = math.ceil(budget_added / _application_opcode_budget)
+    innerTxnFeeMicroAlgos = (
+        inner_txn_count * algosdk.constants.min_txn_fee + algosdk.constants.min_txn_fee
+    )
+    dryrun_starting_budget = 70_000  # https://github.com/algorand/go-algorand/blob/3a5e5847bebc21bfcae1f5fe056a78067b16c781/daemon/algod/api/server/v2/dryrun.go#L408
+
+    @Blackbox(input_types=[])
+    @pt.Subroutine(pt.TealType.uint64)
+    def ensure_budget():
+        return pt.Seq(
+            pt.OpUp(mode=pt.OpUpMode.OnCall).ensure_budget(
+                pt.Int(dryrun_starting_budget + budget_added), source
+            ),
+            pt.Global.opcode_budget(),
+        )
+
+    if with_funding:
+        accounts = (
+            [
+                Account(
+                    address=algosdk.logic.get_application_address(
+                        DryRunExecutor.EXISTING_APP_CALL
+                    ),
+                    status="Offline",
+                    amount=innerTxnFeeMicroAlgos,
+                    amount_without_pending_rewards=innerTxnFeeMicroAlgos,
+                )
+            ]
+            if source == pt.OpUpFeeSource.AppAccount
+            else []
+        )
+
+        sp = DryRunExecutor.SUGGESTED_PARAMS
+        sp.fee = (
+            innerTxnFeeMicroAlgos + algosdk.constants.min_txn_fee
+            if pt.OpUpFeeSource.GroupCredit
+            else sp.fee
+        )
+
+        result = _dryrun(ensure_budget, sp, accounts)
+
+        assert result.passed(), result.report()
+        # Since ensure_budget guarantees _at least_ the required budget, the
+        # assertions confirm minimum expected budget added without significantly
+        # overshooting the mark.
+        assert (
+            cast(int, result.budget_added())
+            >= _application_opcode_budget * inner_txn_count
+        )
+        assert (
+            cast(int, result.budget_added())
+            <= _application_opcode_budget * inner_txn_count + _application_opcode_budget
+        )
```

Comment:
@tzaffi I originally wrote the expression as suggested and switched because I think it's easier to interpret failed assertions with the less Pythonic approach.

In any event, I don't feel strongly enough to contest further - I brought in the change via https://github.com/algorand/pyteal/pull/607/commits/7a16034db677828df7fea82bc3e1476c4ffa9da4.